File: leds_unity/Assets/boubles/Bullet.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

namespace Boubles
{
    public class Bullet
    {
        public int characterID;
        public int ledId;
        public float width = 7;
        public int color;
        int numLeds;
        public Shoots shoots;
        public float dir;
        float timer;
        public List<int> leds;
        public bool isOn;
        float pos;
        float speed = 100f;

        public void Init(Shoots shoots, int characterID, int numLeds, int ledId, int _color)
        {
            this.ledId = ledId;
            leds = new List<int>();
            for (int a = 0; a < width; a++)
                leds.Add(0);
            this.shoots = shoots;
            this.numLeds = numLeds;
            Restart(characterID, ledId, _color);
        }
        public void Restart(int characterID, int ledId, int _color)
        {
            this.characterID = characterID;
            this.color = _color;
            this.pos = ledId;
            this.ledId = ledId;
            timer = 0;
            isOn = true;           
        }
        public void OnUpdate(float deltaTime)
        {
            if ( characterID == 1)
            {
                this.pos -= deltaTime * speed;
                if (this.pos < 0)
                {
                    isOn = false; return;
                }
            }
            else
            {
                this.pos += deltaTime * speed;
                if (this.pos > numLeds - 1)
                {
                    isOn = false; return;
                }
            }
           
            this.ledId = (int)pos;
            int thisLed = ledId - (int)(Mathf.Floor(width / 2));
            if (thisLed < 0) thisLed = numLeds - 1 + thisLed;
            else if (thisLed >= numLeds) thisLed = thisLed - numLeds;
            for (int a = 0; a < width; a++)
            {
                if (thisLed < 0) thisLed = numLeds + thisLed;
                else if (thisLed >= numLeds) thisLed = thisLed - numLeds;
                leds[a] = thisLed;
                thisLed++;
            }
        }
        public void Collide()
        {
            this.isOn = false;
        }


    }

}

File: micropython/bubbles/bullet.py
class Bullet:
    def Init(self, shoots, characterID, numLeds, ledId, color):
        self.characterID = characterID
        self.ledId = ledId
        self.width = 7
        self.color = color
        self.numLeds = numLeds
        self.shoots = shoots
        self.dir = 1 
        self.timer = 0
        self.leds = [0] * self.width
        self.isOn = True
        self.pos = ledId
        self.speed = 100.0

    def Restart(self, characterID, ledId, color):
        self.characterID = characterID
        self.color = color
        self.pos = ledId
        self.ledId = ledId
        self.timer = 0
        self.isOn = True

    def OnUpdate(self, deltaTime):
        
        w = self.width
        n = self.numLeds
        
        if self.characterID == 1:
            self.pos -= deltaTime * self.speed
            if self.pos < 0:
                self.isOn = False
                return
        else:
            self.pos += deltaTime * self.speed
            if self.pos > n - 1:
                self.isOn = False
                return

        self.ledId = int(self.pos)
        thisLed = self.ledId - int(w / 2)
        for i in range(w):
            if thisLed < 0:
                thisLed =  + thisLed
            elif thisLed >= n:
                thisLed = thisLed - n
            self.leds[i] = thisLed
            thisLed += 1

    def Collide(self):
        self.isOn = False